pkg/sharing/secret_reconciler.go

```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		return fmt.Errorf("Watching secrets: %s", err)
+	}
+
+	return controller.Watch(&source.Kind{Type: &sgv1alpha1.SecretExport{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.mapSecretExportToSecret),
+	})
+}
+
+func (r *SecretReconciler) mapSecretExportToSecret(a handler.MapObject) []reconcile.Request {
+	// TODO expensive call on every secret export update (no cached client used, etc)
+	secretList, err := r.coreClient.CoreV1().Secrets("").List(metav1.ListOptions{})
+	if err != nil {
+		r.log.Error(err, "Failed fetching list of all secrets")
+		// TODO what should we really do here?
+		return nil
+	}
+
+	var result []reconcile.Request
+	for _, secret := range secretList.Items {
+		result = append(result, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      secret.Name,
+				Namespace: secret.Namespace,
+			},
+		})
+	}
+	return result
+}
+
+func (r *SecretReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", request)
+
+	secret, err := r.coreClient.CoreV1().Secrets(request.Namespace).Get(request.Name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if secret.DeletionTimestamp != nil {
+		// Nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	return r.reconcile(secret, secret.DeepCopy(), log)
+}
+
+func (r *SecretReconciler) reconcile(secret, originalSecret *corev1.Secret, log logr.Logger) (reconcile.Result, error) {
+	const (
+		imagePullSecretAnnKey = "secretgen.carvel.dev/image-pull-secret"
+	)
+
+	if _, found := secret.Annotations[imagePullSecretAnnKey]; !found {
+		return reconcile.Result{}, nil
+	}
+
+	log.Info("Reconciling")
+
+	// Note that "type" is immutable on a secret
+	if secret.Type != corev1.SecretTypeDockerConfigJson {
+		status := SecretStatus{
+			Conditions: []sgv1alpha1.Condition{{
+				Type:    sgv1alpha1.ReconcileFailed,
+				Status:  corev1.ConditionTrue,
+				Message: "Expected secret to have corev1.SecretTypeDockerConfigJson but did not",
+			}},
+		}
+		return r.updateSecret(secret, status, originalSecret)
+	}
+
+	matcher := SecretMatcher{Namespace: secret.Namespace, SecretType: secret.Type}
+	secrets := r.secretExports.MatchedSecretsForImport(matcher)
+
+	newData, err := NewCombinedDockerConfigJSON(secrets)
+	if err != nil {
+		return reconcile.Result{RequeueAfter: 3 * time.Second}, err
+	}
+
+	secret.Data = newData
+
+	status := SecretStatus{
+		Conditions: []sgv1alpha1.Condition{{
+			Type:   sgv1alpha1.ReconcileSucceeded,
+			Status: corev1.ConditionTrue,
+		}},
+		SecretNames: r.statusSecretNames(secrets),
+	}
+
+	return r.updateSecret(secret, status, originalSecret)
+}
+
+func (r *SecretReconciler) updateSecret(secret *corev1.Secret, status SecretStatus,
+	originalSecret *corev1.Secret) (reconcile.Result, error) {
+
+	const (
```

Comment:
given that we're keeping our consts hyperlocal,  is it reasonable to use the single line form rather than the paren-blocks?

Reply:
this is more of my own stylistic choice -- i always use var and const multiline declaration.